File: frontend/lib/creature.ts
export type CreatureStats = {
  hp: number;
  atk: number;
  def: number;
  speed: number;
  magic: number;
};

export type ProviderStatus = {
  provider: string;
  mode: string;
  modelId?: string | null;
  status?: string;
};

export type Creature = {
  id: string;
  name: string;
  description: string;
  element: string;
  archetype: string;
  rarity: string;
  stats: CreatureStats;
  abilities: string[];
  weaknesses: string[];
  visualUrl: string;
  visualGradient: string;
  providerStatus: ProviderStatus;
  rawExtraction: Record<string, unknown>;
  latencyMs: number;
  fallback: boolean;
};

export type BattleTurn = {
  turn: number;
  attackerId: string;
  attackerName: string;
  defenderId: string;
  defenderName: string;
  move: string;
  damage: number;
  modifier: string;
  leftHp: number;
  rightHp: number;
};

export type BattleResult = {
  left: Creature;
  right: Creature;
  turns: BattleTurn[];
  winnerId: string;
  winnerName: string;
  finalHp: Record<string, number>;
};

export const samplePrompts = [
  "a glass phoenix that reflects spells but shatters under thunder",
  "a thunder golem with copper veins and storm fists",
  "a moss vampire blooming in a moonlit crypt",
  "a clockwork basilisk with ticking brass eyes",
  "a mirror kraken that copies enemy attacks from the abyss"
];

File: backend/app/models/creature.py
from typing import Any

from pydantic import BaseModel, Field


class CreatureStats(BaseModel):
    hp: int = Field(gt=0)
    atk: int = Field(gt=0)
    def_: int = Field(alias="def", gt=0)
    speed: int = Field(gt=0)
    magic: int = Field(gt=0)

    model_config = {"populate_by_name": True}


class ProviderStatus(BaseModel):
    provider: str
    mode: str
    modelId: str | None = None
    status: str = "ready"


class Creature(BaseModel):
    id: str
    name: str
    description: str
    element: str
    archetype: str
    rarity: str
    stats: CreatureStats
    abilities: list[str]
    weaknesses: list[str]
    visualUrl: str
    visualGradient: str
    providerStatus: ProviderStatus
    rawExtraction: dict[str, Any]
    latencyMs: int
    fallback: bool


class CreatureExtractRequest(BaseModel):
    description: str


class BattleRequest(BaseModel):
    leftId: str | None = None
    rightId: str | None = None
    left: Creature | None = None
    right: Creature | None = None


class BattleTurn(BaseModel):
    turn: int
    attackerId: str
    attackerName: str
    defenderId: str
    defenderName: str
    move: str
    damage: int
    modifier: str
    leftHp: int
    rightHp: int


class BattleResponse(BaseModel):
    left: Creature
    right: Creature
    turns: list[BattleTurn]
    winnerId: str
    winnerName: str
    finalHp: dict[str, int]
